traits/trait_list_object.py

```diff
@@ -805,24 +523,197 @@ def notifier(self, trait_list, index, removed, added):
         items_event = self.trait.items_event()
         object.trait_items_event(self.name_items, event, items_event)
 
+    # -- list interface -------------------------------------------------------
+
+    def __delitem__(self, key):
+        """ Delete self[key].
+
+        Parameters
+        ----------
+        key : integer or slice
+            Index of the element(s) to be deleted.
+        """
+        removed_count = len(self[key]) if isinstance(key, slice) else 1
+        self._validate_length(len(self) - removed_count)
+        super().__delitem__(key)
+
+    def __iadd__(self, value):
+        """ Implement self += value.
+
+        Parameters
+        ----------
+        value : iterable
+            The items to be added.
+
+        Returns
+        -------
+        self : LengthConstrainedTraitList
+            The modified list.
+        """
+
+        # Convert input to a concrete list for length-checking purposes.
+        value = list(value)
+        self._validate_length(len(self) + len(value))
+        return super().__iadd__(value)
+
+    def __imul__(self, value):
+        """ Implement self *= value.
+
+        Parameters
+        ----------
+        value : integer
+            The multiplier.
+
+        Returns
+        -------
+        self : LengthConstrainedTraitList
+            The modified list.
+        """
+
+        self._validate_length(len(self) * value)
```

Comment:
Nit: If the value is negative, the error message is going to be a bit confusing. Something like: `The 'at_most_five' trait of a type instance must be a list of at most 5 items which are an integer, but you attempted to change its length to -40 elements.`

(One gets a different TypeError message for when value is not integer-like and cannot be compared with an integer, but that's perhaps too nitpicky that I'd not worry much about.)

Reply:
Aargh. It's worse: we'll fail validation in this case, because the negative new length is below the default minimum length of 0. Will fix and add a regression test.

Yes, not worrying about things that can't be converted to or compared to an integer: we should just let the usual Python exception propagate in that case.

Reply:
Fixed, and test added.

Reply:
> Aargh. It's worse: we'll fail validation in this case, because the negative new length is below the default minimum length of 0. Will fix and add a regression test.

Yes, good point. Thanks for fixing that too.